### DISCUSSION TITLE
The reason of not supporting arm Linux.

### DISCUSSION THREAD

**Discussion by leaf-potato** [GreptimeTeam/greptimedb#4473] [Show and tell]
The official website document states that GreptimeDB only supports amd64 Linux. As far as I know, rust supports cross-platform and GreptimeDB can run on Mac M1. When I compile GreptimeDB in the Ubuntu virtual machine installed on Mac M1, the compilation fails. I am curious about the reason.
<img width="1463" alt="image" src="https://github.com/user-attachments/assets/3cadc650-8ae5-4b6d-a184-49a690315d10">

**Comment by fengys1996** [comment]
Try to manually delete these [two lines](https://github.com/GreptimeTeam/greptimedb/blob/main/.cargo/config.toml#L1-L2), and re_compile.

**Comment by MichaelScofield** [comment]
GreptimeDB is definitely can be built on ubuntu, both arm64 and amd64. We know it because we release the binaries. The documents are wrong, would you kind enough to submit a PR? You can paste the compile errors here, let's try to figure out the reason.

**Reply by evenyag** [reply]
We provide pre-built arm64 binaries for Linux.
https://greptime.com/download

The document page:
https://github.com/GreptimeTeam/docs/blob/main/docs/v0.9/en/contributor-guide/getting-started.md
https://github.com/GreptimeTeam/docs/blob/main/docs/nightly/en/contributor-guide/getting-started.md

PRs are welcome!

**Reply by leaf-potato** [reply]
> GreptimeDB is definitely can be built on ubuntu, both arm64 and amd64. We know it because we release the binaries. The documents are wrong, would you kind enough to submit a PR? You can paste the compile errors here, let's try to figure out the reason.

The compile error seem to: 
```
error: linking with `aarch64-linux-gnu-gcc` failed: exit status: 1
  |
  = note: LC_ALL="C" PATH="/home/leaf-potato/.rustup/toolchains/nightly-2024-04-20-aarch64-unknown-linux-gnu/lib/rustlib/aarch64-unknown-linux-gnu/bin:/home/leaf-potato/.cargo/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/
bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin" VSLANG="1033" "aarch64-linux-gnu-gcc" "/tmp/rustcqN9NOb/symbols.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.107x2451esdmbmkp.rc
gu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.10khbg1qibzth0ll.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1137xg0wz5d416i0.rcgu.o
" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1141jzz5qvy7ei8m.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.117b0ziuueioisib.rcgu.o" "/
home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.16adgype15n31ufn.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.16occytz0eh18ovj.rcgu.o" "/home
/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.178kvr549wu0cyqa.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.18yzhk1bfy80kro5.rcgu.o" "/home/lea
f-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.199w8cahuuk0q234.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1c3cgk1hulc66xbs.rcgu.o" "/home/leaf-po
tato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1cut2pa5m44trwob.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1cz3zeerrd5hhpt2.rcgu.o" "/home/leaf-potato
/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1gahjg75nd3fghb4.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1hmzvgawl52fn721.rcgu.o" "/home/leaf-potato/Ope
n-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1i6k6ktcilywn4tt.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1ia22a64tbgt8sd3.rcgu.o" "/home/leaf-potato/Open-So
urce/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1ikbksrudasnx09q.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1m0ejxn17ewrzreh.rcgu.o" "/home/leaf-potato/Open-Source
/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1o4k73hnwknkm7ar.rcgu.o" "/home/leaf-potato/Open-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0.1oaz5uettu77j1rt.rcgu.o" "/home/leaf-potato/Open-Source/gre

...

n-linux-gnu/lib/librustc_std_workspace_core-4ce1a30d69fd6755.rlib" "/home/leaf-potato/.rustup/toolchains/nightly-2024-04-20-aarch64-unknown-linux-gnu/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcore-bdfe3801ac589305.rlib" "/h
ome/leaf-potato/.rustup/toolchains/nightly-2024-04-20-aarch64-unknown-linux-gnu/lib/rustlib/aarch64-unknown-linux-gnu/lib/libcompiler_builtins-65cc36878682e155.rlib" "-Wl,-Bdynamic" "-lc" "-lz" "-lgcc_s" "-lutil" "-lrt" "-lpth
read" "-lm" "-ldl" "-lc" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "/home/leaf-potato/.rustup/toolchains/nightly-2024-04-20-aarch64-unknown-linux-gnu/lib/rustlib/aarch64-unknown-linux-gnu/lib" "-o" "/home/leaf-potato/Open
-Source/greptimedb/target/debug/deps/greptime-716ed39a77b4b9c0" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs"
  = note: collect2: fatal error: ld terminated with signal 9 [Killed]
          compilation terminated.


error: could not compile `cmd` (bin "greptime") due to 1 previous error

```

**Reply by MichaelScofield** [reply]
My quick guesses are ld got killed by OOM. Try cargo build with "-j 4".

**Reply by fengys1996** [reply]
You can try to manually delete these [two lines](https://github.com/GreptimeTeam/greptimedb/blob/main/.cargo/config.toml#L1-L2), and re_compile.

**Reply by leaf-potato** [reply]
> You can try to manually delete these [two lines](https://github.com/GreptimeTeam/greptimedb/blob/main/.cargo/config.toml#L1-L2), and re_compile.

Maybe not. The same problem occurred.

**Reply by fengys1996** [reply]
Try run `cargo clean` before build.

It seems that the cause of the build error is that an incorrect linker is specified.
`aarch64-linux-gnu-gcc` is a cross compiler for ARM64 target. If specify linker as `aarch64-linux-gnu-gcc` on the arm linux platform, it obviously cannot be found and go wrong.

**Reply by leaf-potato** [reply]
`aarch64-linux-gnu-gcc` is installed in the system.
<img width="522" alt="image" src="https://github.com/user-attachments/assets/cf9e3a8b-afc4-4ba0-9a2e-11bd1143befe">

**Reply by leaf-potato** [reply]
> My quick guesses are ld got killed by OOM. Try cargo build with "-j 4".

I found the reason! The linux installed in the virtual machine only has 8g of memory allocated and the program was killed at the linking stage. I changed the configuration to use the `mold` linker with no problems.

The solution was found in this [discussion](https://github.com/orgs/GreptimeTeam/discussions/4275): 
> [waynexia](https://github.com/waynexia): 确保你有足够的内存。或者尝试使用 `mold`，它在链接阶段需要更少的内存。

**Reply by leaf-potato** [reply]
> We provide pre-built arm64 binaries for Linux. https://greptime.com/download
> 
> The document page: https://github.com/GreptimeTeam/docs/blob/main/docs/v0.9/en/contributor-guide/getting-started.md https://github.com/GreptimeTeam/docs/blob/main/docs/nightly/en/contributor-guide/getting-started.md
> 
> PRs are welcome!

The PR of docs will be submitted later.

**Reply by fengys1996** [reply]
> I found the reason! 

Great!

> aarch64-linux-gnu-gcc is installed in the system.

It looks like the Ubuntu installed on the Mac M1 is an x86 architecture. :sweat_smile: I thought it was an arm64 architecture before.

**Comment by leaf-potato** [comment]
The PR: https://github.com/GreptimeTeam/docs/pull/1118/files
